compiler/nir/gcc_extended_asm.nim

```diff
@@ -0,0 +1,379 @@
+## GCC Extended asm stategments nodes that produces from NIR.
+## It generates iteratively, so parsing doesn't take long
+
+## Asm stategment structure:
+## ```nim
+## Asm {
+##   AsmTemplate {
+##     Some asm code
+##     SymUse nimInlineVar # `a`
+##     Some asm code
+##   }
+##   AsmOutputOperand {
+##     # [asmSymbolicName] constraint (nimVariableName)
+##     AsmInjectExpr {symUse nimVariableName} # for output it have only one sym (lvalue)
+##     asmSymbolicName # default: ""
+##     constraint
+##   }
+##   AsmInputOperand {
+##     # [asmSymbolicName] constraint (nimExpr)
+##     AsmInjectExpr {symUse nimVariableName} # (rvalue)
+##     asmSymbolicName # default: ""
+##     constraint
+## }
+##  AsmClobber {
+##    "clobber"
+## }
+##  AsmGotoLabel {
+##    "label" 
+## }
+## ```
+
+# It can be useful for better asm analysis and 
+# easy to use in all nim targets.
+
+import nirinsts, nirtypes
+import std / assertions
+import .. / ic / bitabs
+
+type
+  Det = enum
+    AsmTemplate
+    SymbolicName
+    InjectExpr
+    Constraint
+    Clobber
+    GotoLabel
+    Delimiter
+  
+  AsmValKind = enum
+    StrVal
+    # SymVal
+    NodeVal
+    EmptyVal
+
+  AsmVal = object
+    case kind: AsmValKind
+    of StrVal:
+      s: string
+    # of SymVal:
+    #   sym: SymId
+    of NodeVal:
+      n: NodePos
+    of EmptyVal:
+      discard
+
+  AsmToken = tuple[sec: int, val: AsmVal, det: Det]
+
+  AsmNodeKind* = enum
+    AsmTemplate
+    AsmOutputOperand
+    AsmInputOperand
+    AsmClobber
+    AsmGotoLabel
+
+    AsmInjectExpr
+    AsmStrVal
+
+  GccAsmNode* = ref object
```

Comment:
That's not how we write ASTs in Nim 2023. We now use "packed trees". The files you have touched can guide you.